Add Packets-per-second benchmark

dPhanekham opened the pull request:
Add a packets-per-second benchmark called netperf_aggregate (this can be changed if needed) which uses netperf UDP_RR between a client VM and two server VMs. Done with a slightly modified version of a script that can be found in https://github.com/HewlettPackard/netperf/blob/master/doc/examples/runemomniaggdemo.sh.

This also required me to change the netperf installation to a slightly newer version so that it would compile with the --enable-demo flag with no errors. I have tested to original netperf benchmark also to assure it still works with no issues.

s-deitz commented on perfkitbenchmarker/linux_packages/netperf.py:
```diff
@@ -31,46 +30,45 @@
-  """Copy the tar file for installation.
+  # vm.PushDataFile('netperf.patch', NETLIB_PATCH)
+  # vm.RemoteCommand('cd %s && patch -p2 < netperf.patch' %
+  #                   NETPERF_SRC_DIR)
```
Remove commented code?

s-deitz commented on perfkitbenchmarker/linux_packages/texinfo.py:
```diff
@@ -0,0 +1,31 @@
+"""Module containing build tools installation and cleanup functions."""
+
+
+def _Install(vm):
```
I don't believe this function is used or necessary.

s-deitz commented on perfkitbenchmarker/linux_packages/python_rrdtool.py:
```diff
@@ -0,0 +1,31 @@
+"""Module containing build tools installation and cleanup functions."""
+
+
+def _Install(vm):
```
I don't believe this function is necessary.

s-deitz commented on perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py:
```diff
@@ -0,0 +1,305 @@
+# ./configure --enable-burst --enable-demo --enable-histogram
+# make  # sudo make install if you prefer
+  vm.Install('texinfo')
+  vm.Install('python_rrdtool')
```
Why are texinfo and python_rrdtool necessary?

s-deitz commented on perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py:
```diff
@@ -0,0 +1,305 @@
+        required to run the benchmark.
+  """
+  vms = benchmark_spec.vms
+  vms[1].RemoteCommand('sudo killall netserver')
```
Kill on the second server VM too?

s-deitz commented on perfkitbenchmarker/linux_packages/netperf.py:
```diff
@@ -31,46 +30,45 @@
+def _LoadNetperf(vm):
+  vm.Install('curl')
+  vm.RemoteCommand('cd %s && git clone %s %s' % (
+      INSTALL_DIR, NETPERF_GIT, NETPERF_DIR))
```
Is git clone getting a later version than we were getting with curl?

s-deitz commented on perfkitbenchmarker/linux_packages/netperf.py:
```diff
@@ -31,46 +30,45 @@
+                   ' find_max_burst.sh'
+                   % (NETPERF_EXAMPLE_DIR))
 
-  Tries local data directory first, then NET_PERF_URL
```
Why are we removing this functionality to get a local netperf tar file?

dPhanekham commented on perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py:
```diff
@@ -0,0 +1,305 @@
+  vm_groups:
+    vm_1:
+      vm_spec: *default_single_core
+    vm_2:
```
changed to use client and server group with arbitrary number of server vms

dPhanekham commented on perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py:
```diff
@@ -0,0 +1,305 @@
+    vm.ApplySysctlPersistent({
+        'net.ipv4.tcp_keepalive_time': 60,
+        'net.ipv4.tcp_keepalive_intvl': 60,
+    })
```
moved this to the netperf package. removed from this benchmark and netperf benchmark

dPhanekham commented on perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py:
```diff
@@ -0,0 +1,305 @@
+  stdout_ascii = stdout.encode("ascii")
+
+  for line in stdout_ascii.splitlines():
+    print(line)
```
removed debugging statements

dPhanekham commented:
I went back and changed quite a few things:

name changed to netperf_pps

changed it to use client and server vm groups, with an arbitrary number of servers

Instead of having a copy of the pps script in perfkitbenchmarker, I use the version already in netperf. A small patch file was already being used to make a few changes to netperf, so I added the changes I needed to make to the pps script to that patch file, as well as a small fix that needed to be made so that it would compile with the enable-demo flag.

dPhanekham commented:
Yes. It can be run with the command ./pkb.py --benchmarks=netperf_pps
I can also provide a .yaml file and or the full output of the run if requested

dPhanekham commented:
Here is a .yaml file for it. I can also add this or something similar as an example to the config directory

netperf_pps:
  description: test packets per second performance
  vm_groups:
    servers:
      vm_spec:
        GCP:
          zone: us-east4-a
          machine_type: n1-standard-2
      vm_count: 4
    client:
      vm_spec:
        GCP:
          zone: us-east4-a
          machine_type: n1-standard-2
